controllers/storagecluster/storageclient.go

```diff
@@ -0,0 +1,63 @@
+package storagecluster
+
+import (
+	"fmt"
+
+	ocsclientv1a1 "github.com/red-hat-storage/ocs-client-operator/api/v1alpha1"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/services"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	cutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	storageClientName = "local-storageclient"
+
+	tokenLifetimeInHours         = 48
+	onboardingPrivateKeyFilePath = "/etc/private-key/key"
+)
+
+type storageClient struct{}
+
+var _ resourceManager = &storageClient{}
+
+func (s *storageClient) ensureCreated(r *StorageClusterReconciler, storagecluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	if !storagecluster.Spec.AllowRemoteStorageConsumers {
+		r.Log.Info("Spec.AllowRemoteStorageConsumers is disabled")
+		return s.ensureDeleted(r, storagecluster)
+	}
+
+	storageClient := &ocsclientv1a1.StorageClient{}
+	storageClient.Name = storageClientName
+	storageClient.Namespace = r.OperatorNamespace
+	_, err := cutil.CreateOrUpdate(r.ctx, r.Client, storageClient, func() error {
+		if storageClient.Status.ConsumerID == "" {
+			token, err := services.GenerateOnboardingToken(tokenLifetimeInHours, onboardingPrivateKeyFilePath)
+			if err != nil {
+				return fmt.Errorf("unable to generate onboarding token: %v", err)
+			}
+			storageClient.Spec.OnboardingTicket = token
+		}
+		storageClient.Spec.StorageProviderEndpoint = storagecluster.Status.StorageProviderEndpoint
```

Comment:
What does `storagecluster.Status.StorageProviderEndpoint` points to? Is it a stable address? a DNS name? 
And who even guarantees that the address is already on the status at this point in the reconcile. most likely it is not, or it is stale (because this PR sets the client reconcile before the API server reconcile)

Reply:
> PR sets the client reconcile before the API server reconcile
- as discussed earlier, no, storageclient reconciles only after API server

The plan where we can get service details from env is really cool and in my testing it seem to have a drawback that the service should exist before the env is set in the pod.

After a bit of search, above is already mentioned in a note in https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables. Nevertheless, as the values are already defined as constants in the same pkg, I'll just use the dns name & port.

Reply:
done.